src/osd/PrimaryLogPG.cc

```diff
@@ -9478,8 +9478,16 @@ void PrimaryLogPG::finish_copyfrom(CopyFromCallback *cb)
     obs.oi.clear_omap_digest();
   }
 
-  obs.oi.truncate_seq = cb->results->truncate_seq;
-  obs.oi.truncate_size = cb->results->truncate_size;
+  unsigned flags = 0;
+  map<hobject_t,CopyOpRef>::iterator p = copy_ops.find(obs.oi.soid);
+  if (p != copy_ops.end()) {
+    CopyOpRef cop = p->second;
+    flags = cop->flags;
+  }
+  if (!(flags & CEPH_OSD_COPY_FROM_FLAG_IGNORE_TRUNCATE_SEQ)) {
```

Comment:
> So this has some utility, but I'm not sure if it actually meets the full needs of the CephFS truncate_seq interactions? @ukernel?
> 
> Even if it does, this would be a lot more general if, rather than IGNORE_TRUNCATE_SEQ, the osd op let you specify a truncate_seq and truncate_size to apply while doing the copy. I don't think that's a lot more difficult?

I'm sure this would require a bit more thought and a better understanding of the consequences of such a change.  Because this would involve modifying the struct ceph_osd_op so that its copy_from field (in the union) would include those extra values:

    struct {
            __le64 snapid;
            __le64 src_version;
            __u8 flags; /* CEPH_OSD_COPY_FROM_FLAG_* */
            /*
             * CEPH_OSD_OP_FLAG_FADVISE_*: fadvise flags
             * for src object, flags for dest object are in
             * ceph_osd_op::flags.
             */
            __le32 src_fadvise_flags;
            __le64 truncate_size;
            __le32 truncate_seq;
    } __attribute__ ((packed)) copy_from;

This would mean dealing with protocol versioning, right?  So, the complexity would be higher, but I've really no clear understanding of what this would imply.

Reply:
Oh shoot. I was hoping that we could follow the usual pattern of just creating a struct copy_from2 that still fits inside the ceph_osd_op union here, but that appears to not be the case because adding these two members does indeed push it past the overall size. :(

In that case we'd better examine carefully the semantics required and see if just keeping the existing truncate_seq works here.

Reply:
we have already encodeed src_name and src_oloc into osd_op.indata. I think we can add more fields to it.  
snapid, src_version, flags. src_fadvise_flags all are about src object. Maybe we can encode all of them into sd_op.indata and just use osd_op.extent

Reply:
@gregsfortytwo @ukernel Have any of you guys had a chance to have a look at the new version of the fix I pushed a few weeks ago?  I followed Zheng's suggestion in adding further fields to the osd_op.indata and I guess it should work.

The only thing I didn't know how to handle is how to handle the feature bits -- I kept an OSD_COPY_FILE_RANGE feature overlapping SERVER_NAUTILUS, but that's probably not what we want, specially because this fix could be backported to older versions.